Source/WebCore/Modules/compression/DecompressionStreamDecoder.cpp

```diff
@@ -147,4 +143,72 @@ ExceptionOr<Vector<uint8_t>> DecompressionStreamDecoder::decompress(const uint8_
 
     return output;
 }
+
+#ifdef __APPLE__
+ExceptionOr<bool> DecompressionStreamDecoder::initializeAppleCompressionFramework() 
+{
+    initailized = true;
+    
+    auto result = compression_stream_init(&stream, COMPRESSION_STREAM_DECODE, COMPRESSION_ZLIB);
+    if (result != COMPRESSION_STATUS_OK)
+        return Exception { TypeError, "Initialization Failed."_s };
+
+    return true;
+}
+
+ExceptionOr<Vector<uint8_t>> DecompressionStreamDecoder::decompressAppleCompressionFramework(const uint8_t* input, const size_t inputLength)
+{
+    size_t index = 0, allocateSize = startingAllocationSize, totalSize = 0;
+    Vector<Vector<uint8_t>>storage;
+    storage.reserveCapacity(128);
+
+    compression_status result;    
+    bool shouldDecompress = true;
+
+    if (!initailized) {
+        auto initializeResult = initializeAppleCompressionFramework();
+        if (initializeResult.hasException())
+            return initializeResult.releaseException();
+    }
+
+    stream.src_ptr = input;
+    stream.src_size = inputLength;
+    
+    while (shouldDecompress) {
+        storage.append(Vector<uint8_t>(allocateSize));
+        totalSize += allocateSize;
+
+        stream.dst_ptr = storage[index].data();
+        stream.dst_size = storage[index].size();
+
+        result = compression_stream_process(&stream, (finish) ? COMPRESSION_STREAM_FINALIZE : 0);
+        
+        if (result == COMPRESSION_STATUS_ERROR)
+            return Exception { TypeError, "Failed to Decode Data."_s };
+
+        if ((result == COMPRESSION_STATUS_END && stream.src_size)
+            || (finish && stream.src_size))
+            return Exception { TypeError, "Extra bytes past the end."_s };
+
+        if (!stream.src_size)
+            shouldDecompress = false;
+        else {
+            index++;
+            if (allocateSize < maxAllocationSize)
+                allocateSize *= 2;
+        }
+    }
+
+    storage.at(index).resize(allocateSize - stream.dst_size);
+    totalSize -= stream.dst_size;
+
+    Vector<uint8_t> output;
+    output.reserveCapacity(totalSize);
+    for (auto& storageElement : storage)
+        output.append(storageElement.data(), storageElement.size());
```

Comment:
You might want to consider using SharedBufferBuilder instead of having a Vector of Vector<uint8_t>, then call extractData().

Reply:
Will investigate this today

Reply:
Thanks for the recommendation for `SharedBufferBuilder`. It simplified quite a bit of the logic here.

I should port it over to CompressionStreamEncoder in another patch.